### DISCUSSION TITLE
Version 2.3.0 breaks with typed properties (Typed property [...] must not be accessed before initialization)

### DISCUSSION THREAD

**Discussion by ludgerey** [livewire/livewire#3141] [7. Bugs (archived)]
### Description

Since version 2.3.0 livewire behaves differently with typed properties. I now get `Typed property App\Http\Livewire\Foo::$bar must not be accessed before initialization` errors on actions.

My components look like this:

```php
class ExemptionsForm extends Component
{
    public ?string $valid_until;

    public function mount($model = null) {
        $this->model = $model;

        if ($model) {
            $this->valid_until = $model->valid_until->format('Y-m-d');
        }
    }

    public function submit()
    {
        $this->validate();

        $attributes = [
            'valid_until' => $this->valid_until, // will break here
        ];

        // [...]
     }
}
```

Executing `wire:click="submit()"` I get the following error:

```
Typed property App\Http\Livewire\ExemptionsForm::$valid_until must not be accessed before initialization
```

Which is thrown here:
```php
trait HandlesActions
{
    public function syncInput($name, $value, $rehash = true)
    {
        $propertyName = $this->beforeFirstDot($name);

        throw_if(
/* --> */   ($this->{$propertyName} instanceof Model || $this->{$propertyName} instanceof EloquentCollection) && $this->missingRuleFor($name),
            new CannotBindToModelDataWithoutValidationRuleException($name, $this::getName())
        );

        $this->callBeforeAndAfterSyncHooks($name, $value, function ($name, $value) use ($propertyName, $rehash) {
            throw_unless(
                $this->propertyIsPublicAndNotDefinedOnBaseClass($propertyName),
                new PublicPropertyNotFoundException($propertyName, $this::getName())
            );

```

### Workaround

If I set `null` on every typed property everything works like before. E.g. `public ?string $valid_until = null;`.

Is this the way to go? Do I need to change my code? If yes, maybe we need to update the docs? Or is it common php knowledge to do that?

Or should livewire handle that somehow?

**Context**
- Livewire version: 2.3.0
- Laravel version: 7.28.4
- Browser: Chrome

**Comment by LintonAchmad** [comment]
Arghhh also getting this issue. The variables need to have default values.

**Comment by uteq** [comment]
I am using the same Livewire version with Laravel 8 and I am not experiencing this problem.

**Comment by Mkennethsmith** [comment]
Experiencing this issue too, but only in my tests. 
Thanks @ludgerey assigning null to the values has helped mitigate the problem

**Comment by uteq** [comment]
It took me some time to understand the actual problem.
It appears that it can happen on submit that the valid_untill is actually not assigned if the model does not exist. This is how it should work.

Another approach than adding null at the attribute level is either adding null to valid_untill on mount whenever model is null. Or at submit, which is what I would do. Because this way you will have the most consistent result in your component.

```php
$attributes = [
    'valid_until' => $this->valid_until ?? null
];
```

?? is the null coalition and will make sure that there will not be an error that the valid_untill until was not initialised.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by basepack** [comment]
REOPEN

Still need the `public ?string $valid_until = null;` with the current `dev-master`.

**Comment by komen205** [comment]
Same for me with `public MODEL $something`

**Comment by basepack** [comment]
REOPEN
